### PR TITLE
Add Bytes.concat

### PR DESCRIPTION
**Pull request by Amxx**
Mentionned in https://github.com/OpenZeppelin/openzeppelin-contracts/pull/5680/files#diff-0399261b2cd954212dda0147b86044380b879ab21e1d1b89c564cb48f3552216R217

See #5680

#### PR Checklist

- [x] Tests
- [x] Documentation
- [x] Changeset entry (run `npx changeset add`)

## Summary by Sourcery

Introduce a new concat function in the Bytes library to merge multiple byte buffers into a single bytes value, with comprehensive tests and documentation.

New Features:
- Add Bytes.concat function to concatenate an array of byte buffers into one bytes object.

Documentation:
- Add NatSpec documentation for the new Bytes.concat function.

Tests:
- Add unit tests for concat covering empty arrays, single-item arrays, and multiple-item arrays.

Chores:
- Add changeset entry for a minor version bump.

### CODE REVIEW

File: contracts/utils/Bytes.sol [hunk 1]
```diff
@@ -128,6 +128,35 @@ library Bytes {
         return buffer;
     }
 
+    /**
+     * @dev Concatenate an array of bytes into a single bytes object.
+     *
+     * For fixed bytes types, we recommend using the solidity built-in `bytes.concat` or (equivalent)
+     *`abi.encodePacked`.
```
Comment (ernestognw) [inline contracts/utils/Bytes.sol]:
```suggestion
     * `abi.encodePacked`.
```

**Review by ernestognw — DISMISSED** [review]

File: test/utils/Bytes.test.js [hunk 2]
```diff
@@ -112,6 +113,38 @@ describe('Bytes', function () {
...
+      await expect(this.mock.$concat(items)).to.eventually.equal(ethers.concat(items));
+    });
+
+    it('multiple (variable length) items', async function () {
+      const items = Array.from({ length: 17 }, generators.bytes);
+      items[5] = generators.bytes.zero;
+      items[7] = generators.bytes.zero;
+      items[9] = generators.bytes.bytes(42);
+      items[10] = generators.bytes.bytes(128);
+      items[11] = generators.bytes.zero;
```
Comment (Amxx) [inline test/utils/Bytes.test.js]:
```suggestion
      const items = [
         generators.bytes.zero,
         generators.bytes.bytes(17),
         generators.bytes.zero,
         generators.bytes.bytes(42),
         generators.bytes.bytes(1),
         generators.bytes.bytes(256),
         generators.bytes.bytes(1024),
         generators.bytes.zero,
         generators.bytes.bytes(7),
         generators.bytes.bytes(15),
         generators.bytes.bytes(63),
         generators.bytes.zero,
         generators.bytes.zero,
      ];
```

File: contracts/utils/Bytes.sol [hunk 3]
```diff
@@ -128,6 +128,37 @@ library Bytes {
...
     }
 
+    /**
+     * @dev Concatenate an array of bytes into a single bytes object.
+     *
+     * For fixed bytes types, we recommend using the solidity built-in `bytes.concat` or (equivalent)
+     * `abi.encodePacked`.
+     *
+     * NOTE: this could be done in assembly with a single loop that expands starting at the FMP, but that would be
+     * significantly less readable. It might be work benchmarking the saving of the full-assembly approach.
```
Comment (arr00) [inline contracts/utils/Bytes.sol]:
```suggestion
     * significantly less readable. It might be worth benchmarking the savings of the full-assembly approach.
```

**Review by arr00 — APPROVED** [review]

File: contracts/utils/Bytes.sol [hunk 4]
```diff
@@ -128,6 +128,37 @@ library Bytes {
...
+     * NOTE: this could be done in assembly with a single loop that expands starting at the FMP, but that would be
+     * significantly less readable. It might be worth benchmarking the savings of the full-assembly approach.
+     */
+    function concat(bytes[] memory buffers) internal pure returns (bytes memory) {
+        uint256 length = 0;
+        for (uint256 i = 0; i < buffers.length; ++i) {
+            length += buffers[i].length;
+        }
+
+        bytes memory result = new bytes(length);
```
Comment (ernestognw) [inline contracts/utils/Bytes.sol:146]: I may be wrong but, will the current version reserve the free memory pointer for the return value and then use the next free pointer? If so, maybe the following would make more sense

```suggestion
    function concat(bytes[] memory buffers) internal pure returns (bytes memory result) {
        uint256 length = 0;
        for (uint256 i = 0; i < buffers.length; ++i) {
            length += buffers[i].length;
        }

        result = new bytes(length);
```

**Review by ernestognw — APPROVED** [review]
Looks good. 

I initially thought about this function to be in `Arrays` mimicking the Javascript's API for [flatten](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat). If you guys think this is a good approach, then it may be worth exploring.

File: contracts/utils/Bytes.sol (hunk 4 above)
Comment (Amxx) [reply contracts/utils/Bytes.sol:146]: I've run tests on remix using this code
```solidity
contract Test {
    function alloc1() public pure returns (bytes memory) {
        return new bytes(64);
    }

    function alloc2() public pure returns (bytes memory result) {
        result = new bytes(64);
    }

    function test1() public pure returns (uint256) {
        uint256 fmpBefore = _getFMP();
        alloc1();
        uint256 fmpAfter = _getFMP();
        return fmpAfter - fmpBefore;
    }

    function test2() public pure returns (uint256) {
        uint256 fmpBefore = _getFMP();
        alloc2();
        uint256 fmpAfter = _getFMP();
        return fmpAfter - fmpBefore;
    }

    function _getFMP() private pure returns (uint256 fmp) {
        assembly ("memory-safe") {
            fmp := mload(0x40)
        }
    }
}
```

`test1` and `test2` both return 96 (32 for the length + 64 for content). So there is no difference in the memory allocation between alloc1() and alloc2().

**Comment by Amxx** [comment]
Javascript's flatten is "Array of Array => Array", which makes sens in the case of javascript because Arrays can contain multiple different types.

IMO, Buffer or Strings would be a better match to solidity's bytes than Array. In javascript, Array.flat does NOT merge strings or buffer, which would mean that this concat function (equivalent to `ethers.concat`) doesn't really match Array.flat.

Also IMO, this "bytes concatenate" function is more bytes specific then it is array specific. I guess that can be disputed, but I'm not sure it would make sens to have a similar function for arrays of uint256, or arrays of bytes32. It would make sens to have something similar for arrays of strings, but I would put that in Strings.sol, not in Array.sol.
